File: A4/einstein.js
"use strict"

// teilloesung, d.h die zwei gehoeren zusammen
function* teil(constraint, strasse) {
	for (const i of _.range(strasse.length)) {
		yield* teilAnStelle(constraint, i, strasse)
	}
}

// 
function* teilAnStelle(constraint, i, strasse) {
	if (klappt(strasse[i], constraint)) {
		const newStrasse = _.cloneDeep(strasse)
		newStrasse[i] = _.assign({}, strasse[i], constraint)
		yield newStrasse
	}
}

// ueberpruefung des constraints fuer das haus
const klappt(haus, constraint) {
	for (const key of Object.keys(constraint)) {
		if (haus[key] && haus[key] !== constraint[key]) {
			return false
		}
	}
	return true
}

// generiere loesung
function* loese(constraints, strasse) {
	if (constraints.length === 0) {
		yield strasse
	} else {
		for (const newStrasse of _.head(constraints)(strasse))
			yield* loese(_.tail(constraints), newStrasse)
	}
}

// unsere fuenf haeuser
const loesung = [{},{},{},{},{}]

// constraints
const einstein = [
	c => teil({person: 'brite', farbe: 'rot'}, c),
	c => teil({person: 'schwede', tier: 'hund'}, c),
	c => teil({person: 'daene', getraenk: 'tee'}, c),
	c => linksVon({farbe: 'gruen'}, {farbe: 'weiss'}, c),
	c => teil({farbe: 'gruen', getraenk: 'kaffee'}, c),
	c => teil({tier: 'vogel', zigarette: 'pallmall'}, c),
	c => mitte({getraenk: 'milch'}, c),
	c => teil({farbe: 'gelb', zigarette: 'dunnhill'}, c),
	c => erster({person: 'norweger'}, c),
	c => neben({zigarette: 'marlboro'}, {tier: 'katze'}, c),
	c => neben({tier: 'pferd'}, {zigarette: 'dunnhill'}, c),
	c => teil({getraenk: 'bier', zigarette: 'winfield'}, c),
	c => neben({person: 'norweger'}, {farbe: 'blau'}, c),
	c => teil({person: 'deutsche', zigarette: 'rothmanns'}, c),
	c => neben({zigarette: 'marlboro'}, {getraenk: 'wasser'}, c),
	c => teil({tier: 'fisch'}, c)
]

// und jetzt mach
console.log(loese(einstein, loesung))

File: A4/einstein.pl
leftto(A,B,[A,B|_]).
leftto(A,B,[_|Tail]):- leftto(A,B,Tail).

nextto(A,B,H):- leftto(A,B,H).
nextto(A,B,H):- leftto(B,A,H).

first(A, [A|_]).

middle(A, [_,_,A,_,_]).

schreibe(Wem,[]):- write('Der '), write(Wem), write(' hat den Fisch.'), !.
schreibe(Wem,[Head|Tail]):- write(Head), nl, schreibe(Wem, Tail).

einstein():- H = [_,_,_,_,_],
	%%% land,farbe,tier,getraenk,zigarette
	member([brite,rot,_,_,_], H),
	member([schwede,_,hund,_,_], H),
	member([daene,_,_,tee,_], H),
	leftto([_,gruen,_,_,_],[_,weiss,_,_,_], H),
	member([_,gruen,_,kaffee,_], H),
	member([_,_,vogel,_,pallmall], H),
	middle([_,_,_,milch,_], H),
	member([_,gelb,_,_,dunhill], H),
	first([norweger,_,_,_,_], H),
	nextto([_,_,_,_,marlboro],[_,_,katze,_,_], H),
	nextto([_,_,pferd,_,_],[_,_,_,_,dunhill], H),
	member([_,_,_,bier,winfield], H),
	nextto([norweger,_,_,_,_],[_,blau,_,_,_], H),
	member([deutsche,_,_,_,rothmanns], H),
	nextto([_,_,_,_,marlboro],[_,_,_,wasser,_], H),
	member([Wem,_,fisch,_,_], H), 
	!, schreibe(Wem,H).
